lib/app_profiler/backend.rb

```diff
@@ -0,0 +1,47 @@
+# frozen_string_literal: true
+
+module AppProfiler
+  class Backend
+    def run(params = {}, &block)
+      raise NotImplementedError
+    end
+
+    def start(params = {})
+      raise NotImplementedError
+    end
+
+    def stop
+      raise NotImplementedError
+    end
+
+    def results
+      raise NotImplementedError
+    end
+
+    def running?
+      raise NotImplementedError
+    end
+
+    class << self
+      def run_lock
+        @run_lock ||= Mutex.new
+      end
+    end
+
+    protected
+
+    def acquire_run_lock
+      self.class.run_lock.try_lock
+    end
+
+    def release_run_lock
+      self.class.run_lock.unlock
+    rescue ThreadError
+      AppProfiler.logger.warn("[AppProfiler] run lock not released as it was never acquired")
+    end
+  end
+
+  autoload :StackprofBackend, "app_profiler/backend/stackprof"
+  autoload :VernierBackend, "app_profiler/backend/vernier"
+  DefaultBackend = AppProfiler::StackprofBackend
```

Comment:
Let's nest these inside of `AppProfiler::Backend` instead of defining them in `AppProfiler` inside of the file that's supposed to only define `AppProfiler::Backend`. We aren't using Zeitwerk yet, but I'd like to remain compatible as much as possible (like how viewer and storage constants are defined).

Also, this triggers the autoload immediately, so I don't think it is necessary to define `DefaultBackend`. We can just set `@profiler_backend ||= Backend::Stackprof`, which will be evaluated when the method is called.